Render header components only once in a roundtrip

pekam opened the pull request:
Instead of applying ComponentRenderer for headers every time the
renderer is changed, apply the current renderer beforeClientResponse so
that only the last renderer is used.

Fix #172

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/vaadin/vaadin-grid-flow/176)
<!-- Reviewable:end -->

gilberto-torrezan commented:
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[src/main/java/com/vaadin/flow/component/grid/AbstractColumn.java, line 93 at r1](https://reviewable.io/reviews/vaadin/vaadin-grid-flow/176#-LBZ_oqDCllm-UJvUtyG:-LBZ_oqDCllm-UJvUtyH:brqzjrv) ([raw file](https://github.com/vaadin/vaadin-grid-flow/blob/ca072139ef76cabfd3107697955b599e9e246989/src/main/java/com/vaadin/flow/component/grid/AbstractColumn.java#L93)):*
> ```Java
>                     .beforeClientResponse(this, context -> renderHeader()));
>         } else {
>             renderHeader();
> ```

What happens if you set a ComponentRenderer and after it set a TemplateRenderer? Looks like the rendering will be applied twice.

Suggestion: add a test with this scenario, and if it is the case that the rendering is applied twice, just remove this if/else and apply `beforeClientResponse` for both cases.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-grid-flow/176)*
<!-- Sent from Reviewable.io -->

pekam commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/main/java/com/vaadin/flow/component/grid/AbstractColumn.java, line 93 at r1](https://reviewable.io/reviews/vaadin/vaadin-grid-flow/176#-LBZ_oqDCllm-UJvUtyG:-LBZgQfv4nAo1DCVYZ5e:b-8sipzf) ([raw file](https://github.com/vaadin/vaadin-grid-flow/blob/ca072139ef76cabfd3107697955b599e9e246989/src/main/java/com/vaadin/flow/component/grid/AbstractColumn.java#L93)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

What happens if you set a ComponentRenderer and after it set a TemplateRenderer? Looks like the rendering will be applied twice.

Suggestion: add a test with this scenario, and if it is the case that the rendering is applied twice, just remove this if/else and apply `beforeClientResponse` for both cases.
</blockquote></details>

Actually the only reason why I added this `instanceof` check here is the unit tests. They broke when I postponed the rendering to `beforeClientResponse`.

Yes, the rendering will be applied twice, but the correct rendering will still be applied last so it's just a bit of extra computation. Before this change the rendering was applied even more often unnecessarily.

So I don't know... I would like to remove the `if instanceof` but I don't know how to fix all my unit tests which expect that the template is in place immediately when setting the header/footer.

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-grid-flow/176)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[src/main/java/com/vaadin/flow/component/grid/AbstractColumn.java, line 93 at r1](https://reviewable.io/reviews/vaadin/vaadin-grid-flow/176#-LBZ_oqDCllm-UJvUtyG:-LBZxEEZ8mYn61puccRH:b-1r1y3r) ([raw file](https://github.com/vaadin/vaadin-grid-flow/blob/ca072139ef76cabfd3107697955b599e9e246989/src/main/java/com/vaadin/flow/component/grid/AbstractColumn.java#L93)):*
<details><summary><i>Previously, pekam (Pekka Maanpää) wrote…</i></summary><blockquote>

Actually the only reason why I added this `instanceof` check here is the unit tests. They broke when I postponed the rendering to `beforeClientResponse`.

Yes, the rendering will be applied twice, but the correct rendering will still be applied last so it's just a bit of extra computation. Before this change the rendering was applied even more often unnecessarily.

So I don't know... I would like to remove the `if instanceof` but I don't know how to fix all my unit tests which expect that the template is in place immediately when setting the header/footer.
</blockquote></details>

You can probably mock the `runWhenAttached` call to run it immediately when called by the test. We have something like that in the `test-root-context` project, and Denis knows more about it.

But what about a test swapping ComponentRenderer and TemplateRenderer? Do we have one test like it already?

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-grid-flow/176)*
<!-- Sent from Reviewable.io -->

pekam commented:
Review status: 1 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

*[src/main/java/com/vaadin/flow/component/grid/AbstractColumn.java, line 93 at r1](https://reviewable.io/reviews/vaadin/vaadin-grid-flow/176#-LBZ_oqDCllm-UJvUtyG:-LB_Qs2CBSi9XupnnBLt:btp2tkx) ([raw file](https://github.com/vaadin/vaadin-grid-flow/blob/ca072139ef76cabfd3107697955b599e9e246989/src/main/java/com/vaadin/flow/component/grid/AbstractColumn.java#L93)):*
<details><summary><i>Previously, gilberto-torrezan (Gilberto Torrezan) wrote…</i></summary><blockquote>

You can probably mock the `runWhenAttached` call to run it immediately when called by the test. We have something like that in the `test-root-context` project, and Denis knows more about it.

But what about a test swapping ComponentRenderer and TemplateRenderer? Do we have one test like it already?
</blockquote></details>

I replaced those tests with ITs and added a test for changing from text header to component header

---


*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-grid-flow/176)*
<!-- Sent from Reviewable.io -->

gilberto-torrezan commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/vaadin/vaadin-grid-flow/176#-:-LB_Z5y99ZlQZ-XyNKhP:bnfp4nl)*
<!-- Sent from Reviewable.io -->
